### PR TITLE
fix: show/hide logic for addToCard, quantity, addToQuoteRequest

### PR DESCRIPTION
**Pull request by Sebastian-Haehnlein**
## PR Type

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring 
- [ ] Build-related changes
- [ ] CI-related changes
- [ ] Documentation content changes
- [ ] application / infrastructure changes
- [ ] Other... Please describe:

## What Is the Current Behavior?

Issue Number: ISREST-951

## What Is the New Behavior?

hide addToCard, quantity, addToQuoteRequest in some cases

added criteria product.salePrice

refactored (cleanup, dont repeat yourself...)

## Does this PR Introduce a Breaking Change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: src/app/core/models/product/product.helper.ts [hunk 1]
```diff
@@ -151,4 +151,13 @@ export class ProductHelper {
       };
     }
   }
+
+  static showAddToCart(product: Product): boolean {
+    return !!(
+      product.inStock &&
+      product.availability &&
+      !ProductHelper.isMasterProduct(product) &&
+      (product.salePrice || ProductHelper.isVariationProduct(product) || ProductHelper.isRetailSet(product))
```
Comment (dhhyi) [inline src/app/core/models/product/product.helper.ts]: what is the logic of this line?

Show if it has a sale price -- ok
Show if it is a variation product or retail set even if it has no price? -- is that even possible?

File: src/app/shared/product/components/product-add-to-basket/product-add-to-basket.component.html [hunk 2]
```diff
@@ -1,4 +1,4 @@
-<div *ngIf="product.inStock && product.availability" class="add-to-cart">
+<div *ngIf="showAddToCart(product)" class="add-to-cart">
```
Comment (dhhyi) [inline src/app/shared/product/components/product-add-to-basket/product-add-to-basket.component.html]: we have a MR in our GitLab (!847), which I would like to see this dependant on

File: src/app/core/models/product/product.helper.ts [hunk 3]
```diff
@@ -151,4 +151,13 @@ export class ProductHelper {
       };
     }
   }
+
+  static showAddToCart(product: Product): boolean {
```
Comment (dhhyi) [inline src/app/core/models/product/product.helper.ts]: The location of this method seems wrong. This is somehow mixing concerns.

File: src/app/core/models/product/product.helper.ts (hunk 1 above)
Comment (Sebastian-Haehnlein) [reply src/app/core/models/product/product.helper.ts]: retailset has other attributes: summedUpSalePrice, minSalePrice
the condition for variation-products is not really necessary (removed)

Comment (dhhyi) [reply src/app/core/models/product/product.helper.ts]: okay :+1:

File: src/app/shared/components/product/product-tile/product-tile.component.html [hunk 4]
```diff
@@ -6,24 +6,24 @@
     <ish-product-label [product]="product"></ish-product-label>
   </div>
 
-  <a *ngIf="configuration.displayName" class="product-title" [routerLink]="product | ishProductRoute: category">{{
+  <a *ngIf="is('displayName')" class="product-title" [routerLink]="product | ishProductRoute: category">{{
     product.name
   }}</a>
 
   <ish-product-rating *ishFeature="'rating'" [product]="product"></ish-product-rating>
```
Comment (dhhyi) [inline src/app/shared/components/product/product-tile/product-tile.component.html]: rating display should also be configurable.

File: src/app/shared/components/product/product-tile/product-tile.component.html [hunk 5]
```diff
@@ -34,20 +34,20 @@
```
Comment (dhhyi) [inline src/app/shared/components/product/product-tile/product-tile.component.html]: this one above could become a switch for `displayVariationCount`

File: src/app/shared/components/product/product-row/product-row.component.html [hunk 6]
```diff
@@ -11,31 +11,31 @@
   <div class="col-9 col-md-10">
     <div class="row">
       <div class="col-md-7 col-xs-12">
-        <a *ngIf="configuration.displayName" class="product-title" [routerLink]="product | ishProductRoute: category">{{
+        <a *ngIf="is('displayName')" class="product-title" [routerLink]="product | ishProductRoute: category">{{
           product.name
         }}</a>
 
         <ish-product-rating *ishFeature="'rating'" [product]="product"></ish-product-rating>
```
Comment (dhhyi) [inline src/app/shared/components/product/product-row/product-row.component.html]: rating display should also be configurable

File: src/app/shared/components/product/product-row/product-row.component.html [hunk 7]
```diff
@@ -46,21 +46,21 @@
...
       <div class="col-12 col-md-5 text-md-right">
         <ish-product-price
-          *ngIf="configuration.displayPrice"
+          *ngIf="is('displayPrice')"
           [product]="product"
           [showInformationalPrice]="true"
         ></ish-product-price>
         <ish-product-inventory
-          *ngIf="configuration.displayInventory && !isMasterProduct(product)"
+          *ngIf="is('displayInventory') && !isMasterProduct(product)"
```
Comment (dhhyi) [inline src/app/shared/components/product/product-row/product-row.component.html]: `!isMasterProduct` should be part of the configuration function

File: src/app/shared/components/product/product-row/product-row.component.html [hunk 8]
```diff
@@ -46,21 +46,21 @@
...
           [product]="product"
         ></ish-product-inventory>
         <ish-product-shipment
-          *ngIf="configuration.displayShipment && !isMasterProduct(product)"
+          *ngIf="is('displayShipment') && !isMasterProduct(product)"
```
Comment (dhhyi) [inline src/app/shared/components/product/product-row/product-row.component.html]: `!isMasterProduct` should be part of the configuration function

File: src/app/shared/components/product/product-add-to-basket/product-add-to-basket.component.ts [hunk 9]
```diff
@@ -83,4 +83,13 @@ export class ProductAddToBasketComponent implements OnInit, OnDestroy {
   ngOnDestroy() {
     this.destroy$.next();
   }
+
+  hasAddToBasketButton() {
```
Comment (dhhyi) [inline src/app/shared/components/product/product-add-to-basket/product-add-to-basket.component.ts]: Why here? Isn't this already decided by parent components?

Comment (Sebastian-Haehnlein) [reply src/app/shared/components/product/product-add-to-basket/product-add-to-basket.component.ts]: not really, ProductAddToBasketComponent is used in six other components and just "tile" and "row" has this condition in the ProductItemContainerConfiguration

File: src/app/shared/components/product/product-tile/product-tile.component.html (hunk 5 above)
Comment (Sebastian-Haehnlein) [reply src/app/shared/components/product/product-tile/product-tile.component.html]: ?

File: src/app/shared/components/product/product-add-to-basket/product-add-to-basket.component.ts (hunk 9 above)
Comment (dhhyi) [reply src/app/shared/components/product/product-add-to-basket/product-add-to-basket.component.ts]: But then the logic is executed twice for product-rows and product-tiles, which is not what we want. Either we do it always from the outside or we do it from inside the component.

It seems that in ISML it is done in the component itself. But then there would be no need to have configuration from the outside.

So we either do it always from the outside (then we'll need a helper method after all) or we only let the add-to-basket component decide when to show itself.

opinions? @shauke

**Comment by shauke** [comment]
Something is not right with the quantity labels in the product listing in row view.
quantity-label-error

**Comment by Sebastian-Haehnlein** [comment]
> Something is not right with the quantity labels in the product listing in row view.
> quantity-label-error

fixed

File: src/app/shared/components/product/product-add-to-basket/product-add-to-basket.component.ts (hunk 9 above)
Comment (shauke) [reply src/app/shared/components/product/product-add-to-basket/product-add-to-basket.component.ts]: The issue whether to show the Add To Cart button or not has two aspects.

It should be configurable at parent components like the product tile, where we might want to use the same product tile component at several places with different rendering styles showing the add to cart button at some places and not at others. So this needs to be for some parent components a display configuration (e.g. for the product tile and product row, but not for product detail or product compare). This is the case in ISML too (see ProductTile.isml).

On the other hand there needs to be logic at the Add To Cart component itself too, that determines whether the add to cart button can be show or not. This logic should be independent from the parents configuration to show the add to cart button. This logic is only based on product properties that are needed to show an add to cart button or not. This is the same in ISML too (see AddToCartInc.isml). And it should be independent from the parent so the Add To Cart component can be used in several places without the need to implement the same logic whether to display it or not in several parent components (compare listing, product tile, product row, product detail, ...).

Comment (dhhyi) [reply src/app/shared/components/product/product-add-to-basket/product-add-to-basket.component.ts]: So as an example: The check for availability (inStock, ...) should be made within the component and disabling the button for master products should be done from the outside?

**Comment by dhhyi** [comment]
Replaced by #403

**Pull request closed without merging**